README.md
# Day Log Savings

[![Version][version-image]][github-url][![Downloads][downloads-image]][npm-url][![JavaScript][javascript-image]][github-url][![License][license-image]][license-url]

## Table Of Contents

- [**About**](#about)
- [**Installation**](#installation)
- [**Write**](#write)
- [**Read**](#read)
- [**Remove**](#remove)
- [**Defaults**](#defaults)

## About

Day Log Savings is a simple, zero dependencies, Node.js logger that lets you log things in organized, day rotating files.<br>
The name is a play on words of daylight savings, which started when this module was first created.

Each day, a new log file is created, that file is then categorised into month and year folders.<br>By default, query's are logged like so: `[<time>] [<prefix>] <input>` but you can customize this by using the functions options or changing the defaults. Those query's are then written into log files and saved at `<project root>/logs/<year>/<month>/<day>.log`. For example: `/logs/2020/11/27.log`.

## Installation

```npm install day-log-savings```

## Test

```npm test```

## Write

Writes an input to the logs, customizable with options.

### Usage

**Function**: `<logger>.write(<input>, [options]);`<br>
**Returns**: The string which was just logged.

**Input {any}**: The input which you want to be logged.<br>
**Options {object}**: {

- **Prefix {string}**: The prefix which appears before the log input, case sensitive. In case of error, prefix is automatically changed to 'ERROR'. Defaults to 'LOG'.
- **Format {object}**: { Change the format of the dates, timestamps and the log message itself. Use `%<option>` to define where within the string you want said option to appear.

    - **Message {string}**: The format in which your input appears in the logs. Options are '%time', '%date', '%prefix' and '%message'. Defaults to '[%time] [%prefix] %message'.
    - **Time {string}**: The format which the timestamps are displayed in. Options are '%hour', '%minute' and '%second'. Defaults to '%hour:%minute:%second'.
    - **Date {string}**: The format which the date are displayed in. Does not change the path to where logs are saved. Options are '%year', '%month' and '%day'. Defaults to '%year/%month/%day'.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}
- **Length {number}**: The maximum length the input can be before being put on a new line. Defaults to '100'.
- **Console {boolean}**: Whether or not to log the query in the console along with the log file. Errors will always be logged. Defaults to 'false'.
- **Stringify {boolean}**: If the input is an instance of 'Object', then JSON.stringify() it . Defaults to 'true'.
- **Stack {boolean}**: If the input is an instance of 'Error', then use the stack property of it. Defaults to 'true'.

}

### Examples

```js
logger.write('Input using the default options.');
// [00:00:00] [LOG] Input using the default options.

logger.write('Has a custom prefix.', { prefix: 'cUsToM' });
// [00:00:00] [cUsToM] Has a custom prefix.

logger.write('Custom date, time and message format.', { format: { message: '[%date %time] [%prefix]: %message', date: '%day/%month/%year', time: '%hour.%minute.%second' } });
// [27/11/2020 00.00.00] [LOG]: Custom date, time and message format.

logger.write('Max first line input length reached.', { length: 1 });
// [00:00:00] [LOG]
// Max first line input length reached.

logger.write("This will be logged in the console and log file.", { console: true });
// Console & Log File:
// [00:00] [LOG] This will be logged in the console and log file.

logger.write(new Error('This error will not be stacked.'), { stack: false });
// [00:00:00] [ERROR] This error will not be stacked.

logger.write({ thisObjectWill: 'not be stringified' }, { stringify: false });
// [00:00:00] [LOG] [object Object]
```

## Read 

Reads and outputs the last x number lines from the bottom of a log file.

### Usage

**Function**: `<logger>.read([options]);`<br>
**Returns**: The last x number of lines of a log file.

**Options {object}**: {

- **Path {string}**: The path, formatted as 'year/month/day', to the log file which you want to read. Defaults to to todays date.
- **Lines {number}**: The number of lines you want to read. Defaults to '15'.
- **Array {boolean}**: Whether you want the output in an array (where one line equal one item) or not. Defaults to 'false'.
- **Blanks {boolean}**: Whether or not to include blank lines in the output, both string and array. Defaults to 'true'.

}

### Examples

```js
logger.read();
// Returns a string containing the last 15 lines of todays log file.

logger.read({ path: '2020/11/27', lines: 5 });
// Returns a string containing the last 5 lines of the 27th of November 2020 log file.

logger.read({ array: true, blanks: false });
// Returns an array containing the last 15 lines of todays log file with all the blank lines removed.
```

## Remove

Deletes a log file.

## Usage

**Function**: `<logger>.remove([path]);`<br>
**Returns**: The path, formatted as 'year/month/day', to the file that was just deleted.

**Path {string}**: The path, formatted as 'year/month/day', to the log file which you want to delete. Defaults to to todays date.

```js
logger.remove();
// Deletes todays log file.

logger.remove('2020/11/27');
// Deletes the 27th of November 2020 log file.
```

## Defaults

Change the defaults for one of the functions that the module has.

### Usage

**Function**: `<logger>.defaults(<method>, [options]);`<br>
**Returns**: The new defaults object of the chosen function.

**Method {string}**: The name of the function you want to change the defaults for. Example: 'write', 'read'.<br>
**Options {object}**: The new defaults which you want to set.

### Examples

```js
logger.defaults('write', { prefix: 'INFO', format: '%date %time %prefix: %message' });
// Changes the default prefix to 'INFO' and the format to '%date %time %prefix: %message'.

logger.defaults('read', { array: true, blanks: false });
// Ensures that the read function returns an array and removes all the blank lines.

logger.defaults('root', { path: `${process.cwd()}/achieve/logs` });
// Changes the default log root from '<project root>/logs' to '<project root>/achieve/logs'.
```

[version-image]: https://img.shields.io/github/package-json/v/ApteryxXYZ/day-log-savings?logo=github
[downloads-image]: https://img.shields.io/npm/dt/day-log-savings?logo=npm
[javascript-image]: https://img.shields.io/github/languages/top/ApteryxXYZ/Day-Log-Savings?logo=github
[license-image]: https://img.shields.io/npm/l/day-log-savings?logo=github

[npm-url]: https://npmjs.com/package/day-log-savings
[license-url]: https://github.com/ApteryxXYZ/Day-Log-Savings/blob/master/LICENSE
[github-url]: https://github.com/ApteryxXYZ/Day-Log-Savings/


CHANGELOG.md
## 0.1.0 2020/09/30
The first release.
- Added write function (writes to a log file).
- Added read function (reads a log file).

### 0.1.1, 0.1.2, 0.1.3 & 0.1.4 2020/09/30
- Removed node-schedule as it was unused.
- Made some small changes.
- Updated README.md.
- Fixed path problem in read function.

### 0.1.5 2020/10/01
- Fixed a few misspellings.
- Updated the README.md.

### 0.1.6 2020/10/02
- Fixed requiring of some modules.

## 0.2.0 2020/10/02
- The read function now reads the entire log file.
- To read the last few lines of a file, use the 'lines' option.

## 0.3.0 2020/10/06
- Added remove function (deletes a log file).

### 0.3.1 2020/10/06
- Updated README.md.

## 0.4.0 2020/10/20
- Logs will now only have time, rather than both date and time.
- To use both date and time use the 'format' option.

## 0.5.0 2020/10/30
- Module no longer uses any other dependencies.
- The read function no longer reads the entire file, but the last 15 lines (by default).

### 0.5.1 2020/10/31
- Fixed remove function removing todays log file even with input.
- Updated README.md.

## 0.6.0 2020/11/23
- Fixed read function reading todays log file even with input.
- Added 'stringify' option to write function, stringify the input if it's an instance of 'Object'.
- Remove function now returns the date/path (2020/11/23) to the file that was just deleted.
- Added more information to errors.
- Updated README.md.

## 0.7.0 2020/11/27
- Added 'defaults' function, change function defaults with a once line.
- Added 'format' option to write function, changes the format of the logs.
- Added 'console' option to  write function, logs input within the console along with log file.
- Added 'blanks' option to read function, if false, removes all the blank lines from the output.
- Moved read function parameter 'path' to within the options object (read({ path: '2020/11/27' })).
- Changed some of the defaults.
- Updated README.md.

### 0.7.1 & 0.7.2 2020/11/27
- Fixed the 'index.d.ts' file.
- Fixed some spelling mistakes.

## 0.8.0 2020/11/30
- Write option 'format' now takes an object containing 'message', 'time' and 'date'.
- Moved extra functions from 'src/index.js' to 'src/\_.js'.
- Added ability to change default file extension via defaults function.
- Increased max first line length default from '100' to '150'.

# 1.0.0 2021/01/01
- Version 1.0.0

### 1.0.1 2021/01/04
- Fixed 'TypeError: Converting circular structure to JSON' when trying to stringify objects with circular references.

### 1.0.2 2021/01/05
- Fixed objects getting stringified twice.

### 1.0.3 & 1.0.4 2021/02/16
- Added proper test.
- Defined node engine in package.json.
- Fixed test folder being in .npmignore.

test/test.js
const assert = require('assert'),
    join = require('path').join;

describe('logger', function () {
    const logger = require('../'),
        package = require('../package.json');

    describe('#defaults', function () {
        it('should set new defaults for a function and return them', function () {
            let newDefaults = logger.defaults('root', { path: join(__dirname, '/logs') });
            assert(logger.DEFAULTS.ROOT, newDefaults);
        })
    })

    describe('#write', function () {
        it('should write to todays log file the inputted value, then return what was logged', function () {
            let loggedValue = logger.write('value', { prefix: 'CUSTOM' });
            assert(loggedValue, logger.read({ lines: 1 }))
        })
    })

    describe('#read', function () {
        it('should read the last lines of the selected log file', function () {
            let readString = logger.read({ lines: 1 }),
                readArray = logger.read({ array: true });
            assert(readString, readArray[readArray.length - 1]);
        })
    })

    describe('#remove', function () {
        it('should delete the selected log file', function () {
            let removedFile = logger.remove(),
                date = new Date();
            assert(removedFile,
                `${date.getFullYear()}/${('0' + (date.getMonth() + 1)).slice(-2)}/${('0' + date.getDate()).slice(-2)}`);
        })
    })

    describe('#name', function () {
        it('should return the name of this node module', function () {
            assert(logger.name, package.name);
        })
    })

    describe('#description', function () {
        it('should return the description of this node module', function () {
            assert(logger.description, package.description);
        })
    })

    describe('#version', function () {
        it('should return the version of this node module', function () {
            assert(logger.version, package.version);
        })
    })
})


test/index.js
var logger;

try {
    logger = require("day-log-savings");
} catch (e) {
    try {
        logger = require(`${process.cwd()}/node_modules/day-log-savings`);
    } catch (e) {
        logger = require("../");
    }
}

logger.write('Input using the default options.');
logger.write('Has a custom prefix.', { prefix: 'cUsToM' });
logger.write('Custom format with date.', { format: '%date %time %prefix: %message' });
logger.write('Max first line input length reached.', { length: 1 });
logger.write("This will be logged in the console and log file.", { console: true });
logger.write(new Error('This error will not be stacked.'), { stack: false });
logger.write({ thisObjectWill: 'not be stringified' }, { stringify: false });

console.log("\nRegular:", logger.read());
console.log("\nArray:", logger.read({ array: true }));
console.log("\nString 5 Lines No Blanks:", logger.read({ lines: 5, blanks: false }));

logger.remove();


src/_.js
const { resolve: rs, sep: s } = require('path'),
    { readSync: re } = require('fs');

function date(f) {
    let n = new Date(),
        y = n.getFullYear(),
        m = ('0' + (n.getMonth() + 1)).slice(-2),
        d = ('0' + n.getDate()).slice(-2);

    return f.replace(/%year/gi, y)
        .replace(/%month/gi, m)
        .replace(/%day/gi, d);
}

function time(f) {
    let n = new Date(),
        h = ('0' + n.getHours()).slice(-2),
        m = ('0' + n.getMinutes()).slice(-2),
        s = ('0' + n.getSeconds()).slice(-2);

    return f.replace(/%hour/gi, h)
        .replace(/%minute/gi, m)
        .replace(/%second/gi, s);
}

function path(D, o = {}) {
    return rs(D.ROOT.path + s + (o.path || date('%year/%month/%day')) + D.ROOT.extension);
}

function options(o, f) {
    if (f == 'write') {
        if (typeof o.prefix != 'string') throw new TypeError('Write o property \'prefix\' must be a string.');
        if (typeof o.format != 'object') throw new TypeError('Write o property \'format\' must be an object.');
        if (typeof o.format.message != 'string') throw new TypeError('Write o property \'format.message\' must be a string.');
        if (typeof o.format.time != 'string') throw new TypeError('Write o property \'format.time\' must be a string.');
        if (typeof o.format.date != 'string') throw new TypeError('Write o property \'format.date\' must be a string.');

        if (typeof o.length != 'number') throw new TypeError('Write o property \'length\' must be a number.');
        if (typeof o.console != 'boolean') throw new TypeError('Write o property \'console\' must be a boolean.');
        if (typeof o.stringify != 'boolean') throw new TypeError('Write o property \'stringify\' must be a boolean.');
        if (typeof o.stack != 'boolean') throw new TypeError('Write o property \'stack\' must be a boolean.');
    } else if (f == 'read') {
        if (typeof o.path != 'string' && o.path) throw new TypeError('Read o property \'path\' must be a string.');
        if (typeof o.array != 'boolean') throw new TypeError('Read o property \'array\' must be a boolean.');
        if (typeof o.lines != 'number') throw new TypeError('Read o property \'lines\' must be a number.');
        if (typeof o.blanks != 'boolean') throw new TypeError('Read o property \'blanks\' must be a boolean.');
    } else if (f == 'root') {
        if (typeof o.path != 'string') throw new TypeError('Root o property \'path\' must be a string.');
        if (typeof o.extension != 'string') throw new TypeError('Write o property \'extension\' must be a string.');
    }
}

function char(f, s, c) {
    let b = Buffer.alloc(1);
    re(f, b, 0, 1, s.size - (c + 1));
    return String.fromCharCode(b[0]);
}

function object(i) {
    return i && typeof i == 'object' && !Array.isArray(i);
}

function merge(t, s) {
    let o = Object.assign({}, t);
    if (object(t) && object(s)) {
        Object.keys(s).forEach(k => {
            if (object(s[k])) {
                if (!(k in t))
                    Object.assign(o, { [k]: s[k] });
                else
                    o[k] = merge(t[k], s[k]);
            } else {
                Object.assign(o, { [k]: s[k] });
            }
        });
    }
    return o;
}

module.exports = { date, time, path, options, char, object, merge };


logger.js
/* CONSTS AND REQUIRES */

const { existsSync, mkdirSync, appendFile, readFileSync, unlinkSync } = require("fs"),
    { readLastLines } = require("read-last-lines-ts"),
    { resolve } = require("path"),
    moment = require("moment");

/* PRIVATE FUNCTIONS */

function check(options, from) {
    if (from === "write") {
        if (options) {
            if (typeof options !== "object") throw new Error("Type of options must be an object.", "DayLog WriteError");
            if (options.prefix && typeof options.prefix !== "string") throw new Error("Type of options.prefix must be a string.", "DayLog WriteError");
            if (options.format && typeof options.format !== "string") throw new Error("Type of options.format must be a string.", "DayLog WriteError");
            if (options.length && typeof options.length !== "number") throw new Error("Type of options.length must be a number.", "DayLog WriteError");
            if (options.stack && typeof options.stack !== "boolean") throw new Error("Type of options.stack must be a boolean.", "DayLog WriteError");
            return { prefix: options.prefix || "auto", format: options.format || "YYYY/MM/DD HH:mm:ss", length: options.length || 100, stack: options.stack || false };
        } else return { prefix: "auto", format: "YYYY/MM/DD HH:mm:ss", length: 100, stack: false };
    } else if (from === "read") {
        if (options) {
            if (typeof options !== "object") throw new Error("Type of options must be an object.", "DayLog ReadError");
            if (options.array && typeof options.array !== "boolean") throw new Error("Type of options.array must be a boolean.", "DayLog ReadError");
            if (options.lines && typeof options.lines !== "number") throw new Error("Type of options.lines must be a number.", "DayLog ReadError");
            return { array: options.array || false, lines: options.lines || 0 };
        } else return { array: false, lines: 0 };
    };
};

function ensure() {
    var root = resolve(process.cwd(), "logs") + "/",
        year = root + moment().format("YYYY") + "/",
        month = year + moment().format("MM") + "/";
    for (dir of [root, year, month]) if (!existsSync(dir)) mkdirSync(dir);
    return month;
};

/* PUBLIC FUNCTIONS */

/**
* Writes something to the logs, customizable with options.
* @param {string} input [Required] The input of which you want to be logged.
* @param {object} options [Optional] Additional options to customize the input.
* @param {string} [options.prefix] The prefix which appears before the log input, case sensitive. Use auto to automatically choose an appropriate prefix. Defaults to 'auto'.
* @param {string} [options.format] The format of the date and time, does not change log path. Uses moments format of formating. Defaults to 'YYYY/MM/DD HH:mm:ss'.
* @param {number} [options.length] The maximum length the input can be before being put on a new line. Defaults to '100'.
* @param {boolean} [options.stack] In case of an error, use stack property of Error. Defaults to 'false'.
* @example
* logger.write("Regular input.");
* // [2020/09/30 00:00:00] [LOG] Regular input.

* logger.write("Custom prefix.", { prefix: "CUSTOM"});
* // [2020/09/30 00:00:00] [CUSTOM] Custom prefix.

* logger.write("A new date and time format.", { format: "HH:mm:ss DD/MM/YYYY" });
* // [00:00:00 30/09/2020] [LOG] A new date and time format.

* logger.write("Max first line length.", { length: 1 });
* // [2020/09/30 00:00:00] [LOG]
* // Max first line length.
*/
function write(input, options) {
    if (!input) throw new Error("Missing required input parameter.", "DayLog WriteError");
    options = check(options, "write");
    var error = input instanceof Error, prefix = options.prefix;
    if (prefix === "auto") prefix = error ? "ERROR" : "LOG";
    if (options.stack && error) input = input.stack || input;
    var time = moment().format(options.format),
        path = ensure() + moment().format("DD") + ".log";
    if (error) input = `\n[${time}] [${prefix}]\n${input}\n\n`;
    else if (input.length < options.length) input = `[${time}] [${prefix}] ${input}\n`;
    else `[${time}] [${prefix}]\n${input}\n`;
    return appendFile(path, input, (err) => { if (err) throw err; }) || "200";
};

/**
* Reads and outputs an entire log file, or using the options, can just output the last number lines.
* @param {string} path [Optional] The path to the file you want to read. Defaults to todays logs.
* @param {object} options [Optional] Additional options to customize the output.
* @param {boolean} [options.array] Whether you want the output in an array or not. Defaults to 'false'.
* @param {number} [options.lines] The number of latest lines you want read, instead of the entire file. Defaults to 'null'.
* @example
* logger.read();
* // All the lines of todays log file.

* logger.read(null, { lines: 100 });
* // Latest 100 lines of todays log file.

* logger.read(null, { array: true });
* // All the lines of todays log file as an array, one item per line.

* loger.read("2020/09/30");
* // All the lines of 30th of the September 2020 log file.
*/
function read(input, options) {
    if (input && typeof input !== "string") throw new Error("Type of input must be a string.");
    var path = resolve(process.cwd(), "logs") + "/" + (input || moment().format("YYYY/MM/DD")) + ".log", lines;
    if (!existsSync(path)) throw new Error(`File at ${path} does not exist.`, "DayLog ReadError");
    options = check(options, "read");
    if (options.lines > 0) lines = readLastLines(path, options.lines).toString("utf8").split("\n").slice(0, -1);
    else lines = readFileSync(path).toString().split("\n").slice(0, -1);
    return options.array ? lines : lines.join("\n");
};

/**
* Deletes a log file.
* @param {string} path [Optional] The path to the file you want to read. Defaults to todays logs.
* @example
* logger.remove();
* // Deletes todays log file.

* loger.remove("2020/09/30");
* // Deletes the 30th of the September 2020 log file.
*/
function remove(input) {
    if (input && typeof input !== "string") throw new Error("Type of input must be a string.");
    var path = resolve(process.cwd(), "logs") + "/" + (input || moment().format("YYYY/MM/DD")) + ".log";
    if (!existsSync(path)) throw Error(`File at ${path} does not exist.`);
    return unlinkSync(path, (err) => { if (err) throw err; }) || "200";
};

/* MODULE EXPORTS */

module.exports = {
    write: write,
    read: read,
    remove: remove
};


src/index.js
const { mkdirSync: mk, appendFileSync: af, existsSync: ex, openSync: os, statSync: ss, closeSync: cs, unlinkSync: ul, } = require('fs'),
    { resolve: rs } = require('path'),
    { inspect: is } = require('util'),
    { name, description, version } = require('../package.json'),
    _ = require('./_'),
    DEFAULTS = {
        WRITE: {
            prefix: 'LOG', format: {
                message: '[%time] [%prefix] %message',
                time: '%hour:%minute:%second', date: '%year/%month/%day'
            },
            length: 150, console: false, stringify: true, stack: true
        },
        READ: { path: undefined, array: false, lines: 15, blanks: true },
        ROOT: { path: `${process.env.PWD}/logs`, extension: '.log' }
    };

function write(input, options = {}) {
    if (!input) throw new Error('Write function is missing required \'input\' parameter.');

    let cp = Object.assign({}, DEFAULTS.WRITE),
        o = _.merge(cp, options);
    _.options(o, 'write');

    let e = input instanceof Error;
    if (o.stack && e) input = input.stack;
    mk(rs(`${DEFAULTS.ROOT.path}/${_.date('%year/%month')}`), { recursive: true });
    if (input instanceof Object && o.stringify && !e) input = is(input);
    let t = _.time(o.format.time),
        d = _.date(o.format.date);

    let q = o.format.message
        .replace(/%time/gi, t)
        .replace(/%date/gi, d)
        .replace(/%prefix/gi, e ? 'ERROR' : o.prefix)
        .replace(/%message/gi, input.length > o.length ? `\n${input}\n` : `${input}\n`);

    if (o.console || e) console.log(q.trim());
    return af(_.path(DEFAULTS), q, (err) => { if (err) throw err }) || q.trim();
}

function read(options = {}) {
    let cp = Object.assign({}, DEFAULTS.READ),
        o = Object.assign(cp, options);
    _.options(o, 'read');

    let p = _.path(DEFAULTS, o);
    if (!ex(p)) throw new Error(`File at '${p}' does not exist.`);

    let f = os(p),
        s = ss(p),
        c = 0, l = 0, ls = '';

    while (c < s.size && l < o.lines + 1) {
        let p = _.char(f, s, c);
        ls = p + ls; c++;
        if (p === '\n' && c > 0) l++;
    }

    cs(f);
    let b = Buffer.from(ls, 'binary').toString().trim(),
        r = o.blanks ? b : b.replace(/\n+/g, '\n');
    return o.array ? r.split(/\n/g) : r;
}

function remove(path) {
    let p = _.path(DEFAULTS, { path });
    if (typeof p !== 'string') throw new TypeError('Read parameter `path` must be a string.');
    if (!ex(p)) throw new Error(`File at '${p}' does not exist.`);
    return ul(p, (err) => { if (err) throw err; }) || p;
}

function defaults(method, options = {}) {
    if (!method || !method instanceof String)
        throw new Error('Defaults function is either missing its \'method\' parameter or what was inputted was not a string.');
    if (!['write', 'remove', 'root'].includes(method))
        throw new Error('Defaults function parameter \'method\' is not a valid method name. Ensure it is one of the following: \'write\', \'read\', \'root\'.')
    if (!options || typeof options !== 'object')
        throw new Error('Defaults function is either missing its \'options\' parameter or what was inputted was not an object.');

    _.options(method, options);
    let m = method.toUpperCase()
    return DEFAULTS[method.toUpperCase()] = _.merge(DEFAULTS[method.toUpperCase()], options);
}

module.exports = new class Logger {
    constructor() {
        this.write = write;
        this.read = read;
        this.remove = remove;
        this.defaults = defaults;

        this.name = name;
        this.description = description;
        this.version = version;
    }

    get DEFAULTS () {
        return DEFAULTS;
    }
};


index.d.ts
declare const logger: {
    /**
     * Changes the defaults that the module uses.
     * @param method The name of the function you want to change the defaults for. Example: 'write', 'read'.
     * @param options The new defaults which you want to set.
     * @returns The new defaults object of the chosen function.
     * @example
     * logger.defaults('write', { prefix: 'INFO', format: { message: '%time: %message' } });
     * logger.defaults('read', { array: true, lines: 25 });
     * logger.defaults('root', { path: `${process.cwd()}/achieve/logs`, extension: '.txt' });
     */
    defaults(method: String, options: Object): Object;

    /**
     * Writes an input to the logs, customizable with options.
     * @param {any} input The input which you want to be logged.
     * @param {object} options Additional options to customize that is logged.
     * @param {object} [options.prefix] The prefix which appears before the log input, case sensitive. In case of error, prefix is automatically changed to 'ERROR'. Defaults to 'LOG'.
     * @param {string} [options.format] Change the format of the dates, timestamps and the log message itself. Use '%<option>' to define where within the string you want said option to appear.
     * @param {string} [options.format.message] The format in which your input appears in the logs. Options are '%time', '%date', '%prefix' and '%message'. Defaults to '[%time] [%prefix] %message'.
     * @param {string} [options.format.time] The format which the timestamps are displayed in. Options are '%hour', '%minute' and '%second'. Defaults to '%hour:%minute:%second'.
     * @param {string} [options.format.date] The format which the date are displayed in. Does not change the path to where logs are saved. Options are '%year', '%month' and '%day'. Defaults to '%year/%month/%day'.
     * @param {number} [options.length] The maximum length the input can be before being put on a new line. Defaults to '100'.
     * @param {boolean} [options.console] Whether or not to log the query in the console along with the log file. Errors will always be logged. Defaults to 'false'.
     * @param {boolean} [options.stringify] If the input is an instance of 'Object', then JSON.stringify() it . Defaults to 'true'.
     * @param {boolean} [options.stack] If the input is an instance of 'Error', then use the stack property of it. Defaults to 'true'.
     * @returns {string} The string which was just logged.
     * @example
     * logger.write('Input using the default options.');
     * logger.write('Has a custom prefix.', { prefix: 'cUsToM' });
     * logger.write('Custom format with date.', { format: { message: '%date %time %prefix: %message' } });
     * logger.write('Custom timestamp format.', { format: { time: '%hour:%minute:%second' } });
     * logger.write('Custom date format.', { format: { date: '%day/%month/%year' } });
     * logger.write('Max first line input length reached.', { length: 1 });
     * logger.write('This will be logged in the console and log file.', { console: true });
     * logger.write({ thisObjectWill: 'not be stringified' }, { stringify: false });
     * logger.write(new Error('This error will not be stacked.'), { stack: false });
     */
    write(input: any, options?: { prefix: String, format: { message: String, time: String, date: String }, length: Number, console: Boolean, stringify: String, stack: String }): String;

    /**
     * Reads and outputs the last customizable number lines from the bottom of a log file.
     * @param {object} options Additional options to choose which file to read and to customize what is outputted.
     * @param {string} [options.path] The path, formatted as 'year/month/day', to the log file which you want to read. Defaults to to todays date.
     * @param {number} [options.lines] The number of lines you want to read. Defaults to '15'.
     * @param {boolean} [options.array] Whether you want the output in an array (where one line equal one item) or not. Defaults to 'false'.
     * @param {boolean} [options.blanks] Whether or not to include blank lines in the output, both string and array. Defaults to 'true'.
     * @returns Either an array or a string of the last lines of a log file.
     * @example
     * logger.read();
     * logger.read({ path: '2020/11/27' });
     * logger.read({ array: true, blanks: false });
     */
    read(options?: { path: String, lines: Number, array: Boolean, blanks: Boolean }): String | Array<String>;

    /**
     * Deletes a log file.
     * @param {string} path The path, formatted as 'year/month/day', to the log file which you want to delete. Defaults to to todays date.
     * @returns {string} The path, formatted as 'year/month/day', to the file that was just deleted.
     * @example
     * logger.remove();
     * logger.remove('2020/11/27');
     */
    remove(path?: String): String;

    /**
     * The name of this module.
     */
    name: String;

    /**
     * The description of this module.
     */
    description: String;

    /**
     * The currently installed version of this module.
     */
    version: String;
}

export = logger;
